Updates to memberships, roles, and permissions handling and validation (#7928)
learningequality/kolibri: docs/backend_architecture/uap/concepts_and_definitions.rst

Users
-----

> The user model is ``FacilityUser``. A

Reviewer: ```suggestion
Kolibri's users are instances of the ``FacilityUser`` model, which derives from Django's ``AbstractBaseUser``. A user
```

The user model is ``FacilityUser``. A
``FacilityUser`` is associated with a particular ``Facility``, and the user's
account and data may be synchronized across multiple devices. A
``DeviceOwner`` account is not associated with a particular ``Facility``, but
is specific to one device, and is never synchronized across multiple devices.

> A ``DeviceOwner`` is like a superuser, and has permissions to modify any data

Reviewer: Separate from this PR, but flagging that `DeviceOwner` is still referenced in other places:

https://kolibri-dev.readthedocs.io/en/develop/backend_architecture/uap/models.html#models

> on her own device, however, normally a ``FacilityUser`` only has permissions for some

Reviewer: ```suggestion
on her own device. However, normally a ``FacilityUser`` only has permissions for some
```

Membership
----------

A ``FacilityUser`` can be marked as a member of a ``Collection`` through a ``Membership`` object.

> Being a member of a Collection, requires first being a member of

Reviewer: ```suggestion
Being a member of a Collection requires first being a member of
```

Reviewer: the capitalization of some words seems a little odd. Maybe either format words as code if referring to the class (`Collection`) or lower-case them if referring to the concept (collection)

Another way in which a ``FacilityUser`` can be associated with a particular
``Collection`` is through a ``Role`` object, which grants the user a role with
respect to the ``Collection`` and all the collections below it. A ``Role``

> object also stores the "kind" of the role (currently, one of "admin",

Reviewer: ```suggestion
object stores the "kind" of the role (currently, one of "admin",
```

Class A". We consider user roles to be "downward-transitive" (meaning if you
have a role for a collection, you also have that role for descendents of that
collection). Thus, in our example, we can say that "Bob is also a coach for
Group Q". Furthermore, as Alice is a member of Class A, we can say that "Bob
is a coach for Alice".

To simplify the role structure, a user can be assigned as an admin, coach, or assignable coach
at the Facility level, but only as a coach at the Classroom level. Users cannot

> be assigned roles at the AdHocGroup/LearnerGroup level.

Reviewer: ```suggestion
A user can be assigned certain roles for different collection types:

* ``Facility`` collections: admin, coach, or assignable coach roles
* ``Classroom`` collections: coach roles 
* ``LearnerGroup`` and ``AdHocGroup`` collections: no roles
```

Is it assignable coach or regular coach for `Classroom`?

Author: Classroom is coach only, assignable coach is only at the facility level.
